lib/Conversion/TosaToXTenNN.cpp

```diff
@@ -0,0 +1,225 @@
+/// constants on the multiplications are power-of-two values and are equal.
+/// We assume the first multiplication represents the quantize meaning 1/scale
+/// and the second multiplication the dequantize or simply scale.
+class FoldMulsToQDQOps : public OpRewritePattern<tosa::MulOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tosa::MulOp mulOp,
+                                PatternRewriter &rewriter) const override {
+
+    // The multiplication should have only a single constant value
+    APFloat dequantizeScaleFactor(0.0);
+    if (!m_ConstantFloat(&dequantizeScaleFactor)
+             .match(mulOp.getOperand(1).getDefiningOp())) {
+      return rewriter.notifyMatchFailure(mulOp.getOperand(1).getLoc(),
+                                         "expected to be a constant.");
+    }
+
+    // We want to make sure we have QDQ operations surrounded by MULs.
+    Operation *mulOperandOp = mulOp->getOperand(0).getDefiningOp();
+    APFloat quantizeScaleFactor(0.0);
+    auto isQDQPattern = m_Op<amd::xten_nn::DequantizeOp>(
+        m_Op<amd::xten_nn::QuantizeOp>(m_Op<tosa::MulOp>(
+            matchers::m_Any(), m_ConstantFloat(&quantizeScaleFactor))));
+    if (mulOperandOp == nullptr || !isQDQPattern.match(mulOperandOp)) {
+      return rewriter.notifyMatchFailure(mulOp->getLoc(),
+                                         "expected mul->q->dq->mul pattern.");
+    }
+    auto dequantizeOp = cast<amd::xten_nn::DequantizeOp>(mulOperandOp);
+
+    // The quantize operation will have it's reciprocal value constant folded.
+    // So we need to calculate one over to get the scale factor back.
+    APFloat recipValue((float)1.0);
+    recipValue.divide(quantizeScaleFactor, APFloat::rmNearestTiesToEven);
+    if (recipValue != dequantizeScaleFactor) {
+      return rewriter.notifyMatchFailure(
+          mulOp.getLoc(), "expected constants of both multiplications around "
+                          "the QDQ to be equal.");
+    }
+
+    // Only power of two values are supported by the QDQ operations
+    std::optional<int32_t> scaleFactor =
+        getLog2Value(dequantizeScaleFactor.convertToFloat());
+    if (!scaleFactor.has_value()) {
+      return rewriter.notifyMatchFailure(
+          mulOp.getLoc(), "constant is not a integer log2 value.");
+    }
+
+    auto quantizeOp = cast<amd::xten_nn::QuantizeOp>(
+        dequantizeOp->getOperand(0).getDefiningOp());
+
+    // Sum the shifts of the quantize, dequantize and update the operations
+    llvm::APInt shiftSum(32, dequantizeOp.getShift(), true);
+    bool overflow = false;
+    llvm::APInt scaleFactorInt(32, scaleFactor.value(), true);
+    shiftSum = shiftSum.sadd_ov(scaleFactorInt, overflow);
+    if (overflow) {
+      return rewriter.notifyMatchFailure(
+          mulOp.getLoc(), "Adding the shifts of the mul and QDQ overflowed.");
+    }
+    quantizeOp.setShift((int32_t)shiftSum.getSExtValue());
```

Comment:
I don't think that you are allowed to modify ops without telling the rewriter. Typical, people just create a new quanitze op, copying most parameters.
Otherwise you need to use (start|cancel|finalize)RootUpdate, see https://mlir.llvm.org/docs/PatternRewriter/#pattern-rewriter